jplag/src/main/java/de/jplag/reporting/reportobject/mapper/ComparisonReportMapper.java

```diff
@@ -0,0 +1,107 @@
+package de.jplag.reporting.reportobject.mapper;
+
+import java.io.BufferedReader;
+import java.io.File;
+import java.io.FileReader;
+import java.io.IOException;
+import java.util.*;
+import java.util.stream.Collectors;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import de.jplag.*;
+import de.jplag.reporting.reportobject.model.ComparisonReport;
+import de.jplag.reporting.reportobject.model.FilesOfSubmission;
+import de.jplag.reporting.reportobject.model.Match;
+
+public class ComparisonReportMapper {
+
+    private static final Logger logger = LoggerFactory.getLogger(ComparisonReportMapper.class);
+    private final HashMap<Long, String> lineLookUpTable = new HashMap<>();
+    private long currentLineIndex;
+
+    /**
+     * Generates detailed ComparisonReport DTO for each comparison in a JPlagResult.
+     * @param jPlagResult The JPlagResult to generate the comparison reports from.
+     * @return A ComparisonReportMapperResult consisting of two DTOs: the list of ComparisonsReport. A ComparisonReport
+     * contains information about a comparison between two submission - including their files. These files are not saved in
+     * plain text though, they are saved as numbers. these numbers are indices to the second DTO of the
+     * ComparisonReportMapperResult, the lineLookUpTable.
+     */
+    public ComparisonReportMapperResult generateComparisonReports(JPlagResult jPlagResult) {
+        List<ComparisonReport> comparisons = new ArrayList<>();
+        int maxNumOfComparisons = jPlagResult.getOptions().getMaximumNumberOfComparisons();
+        jPlagResult.getComparisons(maxNumOfComparisons).forEach(comparison -> comparisons.add( //
+                new ComparisonReport(comparison.getFirstSubmission().getName(), //
+                        comparison.getSecondSubmission().getName(), //
+                        comparison.similarity(), //
+                        getFilesForSubmission(comparison.getFirstSubmission()), //
+                        getFilesForSubmission(comparison.getSecondSubmission()), //
+                        convertMatchesToReportMatches(jPlagResult, comparison) //
+                )));
+        return new ComparisonReportMapperResult(comparisons, lineLookUpTable);
+    }
+
+    private List<FilesOfSubmission> getFilesForSubmission(Submission submission) {
+        return submission.getFiles().stream().map(file -> new FilesOfSubmission(file.getName(), readFileLines(file))).collect(Collectors.toList());
+    }
+
+    private List<Long> readFileLines(File file) {
+        List<Long> lineIndices = new ArrayList<>();
+        try (BufferedReader bufferedReader = new BufferedReader(new FileReader(file))) {
+            String line;
+            while ((line = bufferedReader.readLine()) != null) {
```

Comment:
Why should we not simply store the start/end of a match. Why the string ?

Reply:
Afterwards, we can simply compress the submissions of the students, and maybe more if needed

Reply:
I'm not really sure if I know what you mean.. Are you asking why we are reading the code from the submission files and writing it to the report at all? If so, I do think this is the way to go to enable to the UI to have as little complexity as possible.

Also, we do only store beginning and end of matches (see `ComparisonReportMapper` line 101)

Reply:
I want to include the submission code directly in the jsons.
I'm not sure why the lookup table is needed. 

I think:
a) store the source code to the jsons
b) save the start/end lines of a match

should contain the information we need. Why do we need the indirection (map)?

Reply:
That’s exactly the format I‘ve switched to and I‘m currently implementing: https://github.com/jplag/JPlag/pull/507#issuecomment-1193628178